lib/diaspora_federation/entities/relayable_retraction.rb

```diff
@@ -46,7 +46,76 @@ class RelayableRetraction < Entity
       #   This signature is mandatory only when federation from the subscriber to an upstream
       #   author is done.
       #   @return [String] target author signature
-      property :target_author_signature
+      property :target_author_signature, default: nil
+
+      # Generates XML and updates signatures
+      # @see Entity#to_xml
+      # @return [Nokogiri::XML::Element] root element containing properties as child elements
+      def to_xml
+        entity_xml.tap do |xml|
+          hash = to_h
+          self.class.update_signatures!(hash)
+          xml.at_xpath("target_author_signature").content = hash[:target_author_signature]
+          xml.at_xpath("parent_author_signature").content = hash[:parent_author_signature]
+        end
+      end
+
+      # Adds signatures to a given hash with the keys of the author and the parent
+      # if the signatures are not in the hash yet and if the keys are available.
+      #
+      # @param [Hash] hash hash given for a signing
+      def self.update_signatures!(hash)
+        target_author = DiasporaFederation.callbacks.trigger(
+          :fetch_entity_author_id_by_guid,
+          hash[:target_type],
+          hash[:target_guid]
+        )
+        pkey = DiasporaFederation.callbacks.trigger(:fetch_private_key_by_id, hash[:diaspora_id])
+
+        fill_required_signature(target_author, pkey, hash) unless pkey.nil?
+      end
+
+      def self.fill_required_signature(target_author, pkey, hash)
+        if target_author == hash[:diaspora_id] && hash[:target_author_signature].nil?
+          hash[:target_author_signature] = Signing.sign_with_key(apply_signable_exceptions(hash), pkey)
+        elsif target_author != hash[:diaspora_id] && hash[:parent_author_signature].nil?
+          hash[:parent_author_signature] = Signing.sign_with_key(apply_signable_exceptions(hash), pkey)
+        end
+      end
+      private_class_method :fill_required_signature
+
+      # verifies the signatures (+target_author_signature+ and +parent_author_signature+ if needed)
+      # @param [Hash] data hash with data to verify
+      # @raise [SignatureVerificationFailed] if the signature is not valid or no public key is found
+      def self.verify_signatures(data)
```

Comment:
@annando: Diaspora currently sends `signed_retraction`s for Posts and `relayable_retraction` for Relayables and `retraction` for everything else (Photo).

When do you send a retraction for a person? For persons there is a `account_deletion` message.

What we want to change as first step (hopefully by the next diaspora release in 6 weeks) is, that we send `retraction`s instead of `signed_retraction` for posts. Diaspora is able to parse this correctly. The question is: Can friendica handle a `retraction` for a post? Or should we wait with this change?

@Flaburgan: there are currently many things in the federation code, that aren't needed, but were added in the past. But you can't remove them simply, because you need to stay compatible (thats why they are still there, nobody touched these things since years). I want to clean the federation, to make it less complex, more stable and easier to implement by others (friendica/redmatrix). But we need many iterations for this, because I don't want to break something.